### PR TITLE
chore: [Multi-domain] Break out separate CI tasks to test the driver with experimentalSessionAndOrigin on

### PR DESCRIPTION
**Pull request by mjhenkes**
- Closes 

### User facing changelog

n/a

### Additional details

This pr creates 4 new CI jobs to run driver tests with the `experimentalSessionAndOrigin` feature flag enabled.

They are creatively named: 
 
- `driver-integration-tests-firefox-experimentalSessionAndOrigin`
- `driver-integration-tests-chrome-experimentalSessionAndOrigin`
- `driver-integration-tests-chrome-beta-experimentalSessionAndOrigin`
- `driver-integration-tests-electron-experimentalSessionAndOrigin`

These jobs only run tests in the 'e2e/multi-domain' folder

The standard jobs have be updated to run all tests __not__ in the 'e2e/multi-domain' folder.

#### Before
`driver-integration-tests-chrome` test count was [4748](https://app.circleci.com/pipelines/github/cypress-io/cypress/36529/workflows/2b1a3050-100c-4d12-933e-c346fefe1396/jobs/1463415)

#### After
`driver-integration-tests-chrome` test count is [4419](https://app.circleci.com/pipelines/github/cypress-io/cypress/36670/workflows/3336d37d-af65-4d7b-a23c-b09bcb62bb38/jobs/1470177)

`driver-integration-tests-chrome-experimentalSessionAndOrigin` test count is [329](https://app.circleci.com/pipelines/github/cypress-io/cypress/36670/workflows/3336d37d-af65-4d7b-a23c-b09bcb62bb38/jobs/1470196)

for a total of 4748 (4419 + 329)

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/cypress/integration/commands/navigation_spec.js [hunk 1]
```diff
@@ -2306,32 +2280,6 @@ describe('src/cy/commands/navigation', () => {
         return null
       })
     })
-
-    it('does not wait for stability at the end of the command queue when not stable with experimentalSessionAndOrigin', (done) => {
```
Comment (mjhenkes) [inline packages/driver/cypress/integration/commands/navigation_spec.js:2310]: test moved to the stability test file

File: packages/driver/cypress/integration/commands/navigation_spec.js [hunk 2]
```diff
@@ -1410,22 +1410,9 @@ describe('src/cy/commands/navigation', () => {
...
-          \`  <other commands targeting http://localhost:3501 go here>\`
-          \`})\`\n
-          The new URL is considered a different origin because the following parts of the URL are different:\n
-            > port\n
-          You may only \`cy.visit()\` same-origin URLs within a single test.\n
-          The previous URL you visited was:\n
-            > 'http://localhost:3500'\n
-          You're attempting to visit this URL:\n
-            > 'http://localhost:3501'`)
+          expect(err.message).to.include('`cy.visit()` failed because you are attempting to visit a URL that is of a different origin.')
```
Comment (mschile) [inline packages/driver/cypress/integration/commands/navigation_spec.js:1413]: Why where these changed?

File: packages/driver/cypress/integration/e2e/multi-domain/stability_spec.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        cy.on('command:queue:end', () => {
+          expect(onLoad).not.have.been.called
+          done()
+        })
+
+        cy.on('command:queue:before:end', () => {
+        // force us to become unstable immediately
+        // else the beforeunload event fires at the end
+        // of the tick which is too late
+          cy.isStable(false, 'testing')
```
Comment (mschile) [inline packages/driver/cypress/integration/e2e/multi-domain/stability_spec.ts:66]:
```suggestion
          // force us to become unstable immediately
          // else the beforeunload event fires at the end
          // of the tick which is too late
          cy.isStable(false, 'testing')
```

File: packages/driver/package.json [hunk 4]
```diff
@@ -5,7 +5,9 @@
   "scripts": {
     "clean-deps": "rm -rf node_modules",
     "cypress:open": "node ../../scripts/cypress open",
-    "cypress:run": "node ../../scripts/cypress run",
+    "cypress:run": "node ../../scripts/cypress run --spec \"cypress/integration/*/*\",\"cypress/integration/*/!(multi-domain)/**/*\"",
+    "cypress:open-experimentalSessionAndOrigin": "node ../../scripts/cypress open --config experimentalSessionAndOrigin=true",
+    "cypress:run-experimentalSessionAndOrigin": "node ../../scripts/cypress run --config experimentalSessionAndOrigin=true  --spec \"cypress/integration/e2e/multi-domain/**/*\"",
```
Comment (mschile) [inline packages/driver/package.json:10]: Would it make sense to test all files with the flag on to ensure no one adds anything that will break once we turn on the flag?

**Review by mschile — CHANGES_REQUESTED** [review]

File: packages/driver/cypress/integration/commands/navigation_spec.js (hunk 2 above)
Comment (mjhenkes) [reply packages/driver/cypress/integration/commands/navigation_spec.js:1413]: I reverted these back to how the test was prior to our mulit-domain change, but now that you mention it I could probably validate more. Let me check on it.

File: packages/driver/package.json (hunk 4 above)
Comment (mjhenkes) [reply packages/driver/package.json:10]: The tests that changed in the navigation spec will fail if they're run. I could look at flexing them base don the experimental flag maybe?

**Review by flotwig — APPROVED** [review]
This implementation does not spark joy. We are creating 4 new jobs bringing us to 90 jobs once `10.0-release` merges, and there is a hard limit of 100. Also, it's just... not the most elegant solution :smile: 

@mjhenkes tells me that the target date for x-origin is ~2 months out, so it's not the worst thing to have this temporarily. But if this was for longer-term, I'd prefer to see a neater solution, like using an internal `Cypress.backend()` to signal to the proxy that cy.origin is enabled for these tests.

---

@mjhenkes I see that this doesn't revert most of the changes made to the `driver-integration` test suite in `feature-multidomain`, will there be a future PR to revert those changes?

File: packages/driver/cypress/integration/commands/navigation_spec.js (hunk 1 above)
Comment (mjhenkes) [reply packages/driver/cypress/integration/commands/navigation_spec.js:2310]: Now it's been combined with the test here, so it flexes the test based on the experimental flag

File: packages/driver/package.json [hunk 5]
```diff
@@ -5,7 +5,9 @@
   "scripts": {
     "clean-deps": "rm -rf node_modules",
     "cypress:open": "node ../../scripts/cypress open",
-    "cypress:run": "node ../../scripts/cypress run",
+    "cypress:run": "node ../../scripts/cypress run --spec \"cypress/integration/*/*\",\"cypress/integration/*/!(multi-domain)/**/*\"",
```
[the remark this answers is not in the document]
Comment (AtofStryker) [reply packages/driver/package.json:8]: I think the serialization unit tests outside the `multi-domain` directory will run twice once #20949 makes it way in. They should work in both scenarios with the flag on/off, but one of the runs will be redundant. I don't think it's a big deal, but I figure  I would surface it here.

**Review by AtofStryker — APPROVED** [review]
I think the `e2e cross origin navigation` system tests and some of the other `cy.origin` system needs needs to still be system tests because they may completely fail depending on whether the `experimentalSessionAndOrigin` is on or off, but we can migrate these tests back into integration tests once we have a solution similar to what @flotwig is talking about?

File: circle.yml [hunk 6]
```diff
@@ -1294,6 +1307,44 @@ jobs:
...
+    steps:
+      - run-driver-integration-tests:
+          browser: chrome:beta
+          install-chrome-channel: beta
+          experimentalSessionAndOrigin: true
+
+  driver-integration-tests-firefox-experimentalSessionAndOrigin:
+    <<: *defaults
+    resource_class: medium
+    parallelism: 5
```
Comment (emilyrohrbough) [inline circle.yml:1333]: Should we bump this to 6 or 7 since so many tests for multi-domain since we were concerned with longer CI times?

**Comment by mjhenkes** [comment]
@flotwig 
> @mjhenkes I see that this doesn't revert most of the changes made to the driver-integration test suite in feature-multidomain, will there be a future PR to revert those changes?

I was not planning on reverting changes made to the driver-integration test suite that were not addressed in this PR. The changes made are preparing us to be able to enable this feature flag in the future and in general align with our best practice stance of isolating tests from each other. Additionally, with this setup we are proving that the tests will work with the feature flag both on and off.

**Comment by flotwig** [comment]
> The changes made are preparing us to be able to enable this feature flag in the future and in general align with our best practice stance of isolating tests from each other

@mjhenkes It does change the nature of a lot of assertions though... I don't really think it's a good idea to change a bunch of unrelated tests in a PR. If we're merging the *experimental version* of a feature, why would it need to change tests for non-experimental code? We can always cherry-pick the test changes back in, not like we'd lose much.

Also, it was a conscious choice to not follow "best practices" of using `cy.visit()` in each test for some of the changed tests, a lot of the integration tests are very small and simple and take an unnecessary amount of time to run with a `cy.visit()` that doesn't add additional coverage. I'm not particularly married to the idea of doing it this way, but there is a reason.

**Review by mschile — APPROVED** [review]

File: circle.yml (hunk 6 above)
Comment (mjhenkes) [reply circle.yml:1333]: It's hard to get a direct comparison but with current parallelization the CI job seems to run for roughly the same duration as our existing job. So I don't think we need to bump these right now.

**Comment by mjhenkes** [comment]
@flotwig 

> It does change the nature of a lot of assertions though

Could you point out some examples of this? The changes to enable sessions should in theory largely be visiting for each test instead of visiting once in a before. If we're changing what we're asserting on, we should give that a deeper look.

> It does change the nature of a lot of assertions though... I don't really think it's a good idea to change a bunch of unrelated tests in a PR.

You're right. If we could do it over we would have made these changes in develop then merged them down into our brach. Unfortunately at this point we're running up against a time constraint, and while I don't think it's ideal to have these test changes, I also don't think it should prevent the PR from being merged as we have test case coverage.

> If we're merging the experimental version of a feature, why would it need to change tests for non-experimental code? We can always cherry-pick the test changes back in, not like we'd lose much.

We do need these changes, with this PR we are running all driver tests with `experimentalSessionAndOrigin` enabled as well as a subset (minus the 'multi-domain' folder) with the flag disabled. This gives us excellent coverage and reduces friction to enable the feature flag as any added feature tests will have to pass with the flag in both states. I see this as a feature of this approach.

>Also, it was a conscious choice to not follow "best practices" of using cy.visit() in each test for some of the changed tests, a lot of the integration tests are very small and simple and take an unnecessary amount of time to run with a cy.visit() that doesn't add additional coverage. I'm not particularly married to the idea of doing it this way, but there is a reason.

I pass no judgement on either approach, however, once `cy.sessions` is enabled you will no longer be able to visit once in a before block and run multiple tests against it. This decision was made before my time. If you have strong feelings about this, we should make sure to address them before this feature becomes generally available.  I think bringing up these topics is another feature of this approach 😀 

I also really really wish github would let me respond to general comments so I didn't have to do all this quoting nonsense. 😞

**Pull request merged**